recitation_3/structs-unions.md

```diff
@@ -0,0 +1,281 @@
+
+The layout would look like this:
+
+```
+ +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+ | .len = 3      | (padding)     | .str = (points to "abc")      |
+ +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+ |--------------------- struct MyString -------------------------|
+```
+
+Struct padding occurs because C requires data elements within a struct to be stored in memory with a specific alignment. The alignment requirement depends on the data type and the architecture of the computer, but it is typically a multiple of the size of the largest data type.
+
+In the above example, the largest type is `char *`. Therefore, an extra 4 bytes of space are padded after `int len`. `sizeof(struct MyString)` then would evaluate to 16.
+
+A struct may pad differently depending on the order of elements. A poorly ordered struct may look something like this:
+
+```c
+struct MyStuff {
+    char *str1;
+    int s = len1;
+    char *str2;
+    int s len2;
+};
+
+Here, the two `char *`s are the largest type with 8 bytes. :
+
+```
+ +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+ | .str1                         |  .len1        |  (padding)    | .str2                         |  .len2        | (padding)     |
+ +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+ |-------------------------------------------------------- struct MyStuff -------------------------------------------------------|
+```
+
+However, if we are to reorganize the struct like so:
+
+```c
+struct MyStuff2 {
+    char *str1;
+    char *str2;
+    int len1;
+    int len2;
+};
+```
+The layout would look like this:
+
+```
+ +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+ | .str1                         | .str2                         |  .len1        | .len2         |
+ +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+ |------------------------------------ struct MyStuff2 ---------------------- -------------------|
+```
+
+Try to guess what the size of each struct will be below:
+
+```c
+#include <stdio.h>
+
+struct a {
+    char *x;
+    int y;
+    char z;
+};
+
+struct b {
+    int x;
+    char *y;
+    char z;
+};
+
+struct c {
+    char *p;   
+    char c;
+    short x;
+};
+
+struct d {
+    char w;
+    int x;
+    short y;
+    char z;
+};
+
+
+int main() {
+    printf("%lu\n", sizeof(struct a));
+    printf("%lu\n", sizeof(struct b));
+    printf("%lu\n", sizeof(struct c));
+    printf("%lu\n", sizeof(struct d));
+}
+```
+
+### Pointers to structs
+
+Since structs may get very large in size (e.g., struct Buffer), it can be more
+efficient to refer to them via pointers rather than copy them around by value:
```

Comment:
```suggestion
Since structs may get very large in size (e.g., struct Buffer), it can be more efficient to refer to them via pointers rather than copy them around by value:
```